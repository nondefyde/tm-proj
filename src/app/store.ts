import { Action, configureStore, ThunkAction } from '@reduxjs/toolkit';
import personalityTestReducer from '../features/personality-test/PersonalityTestSlice';

export const store = configureStore({
  reducer: {
    personalityTest: personalityTestReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
