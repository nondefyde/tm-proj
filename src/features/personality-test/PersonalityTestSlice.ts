import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { fetchQuestion } from './question-set/questionsAPI';
import { AnswerType, QuestionType } from './question-set/question';
import { RootState } from '../../app/store';

export interface QuestionSetState {
  questions: QuestionType[];
  current?: QuestionType;
  currentIndex: number;
  answers: Record<string, AnswerType>;
  status: 'idle' | 'loading' | 'failed';
}

const initialState: QuestionSetState = {
  questions: [],
  currentIndex: 0,
  current: undefined,
  answers: {},
  status: 'idle',
};

export const fetchQuestionAsync = createAsyncThunk(
  'questions/fetchQuestions',
  async () => {
    const response = await fetchQuestion();
    // The value we return becomes the `fulfilled` action payload
    return response.data;
  }
);

export const personalityTestSlice = createSlice({
  name: 'personalityTest',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    // Use the PayloadAction type to declare the contents of `action.payload`
    selectNextQuestion: (state, action: PayloadAction<string>) => {
      const index = state.currentIndex + 1
      if (state.questions[index]) {
        state.current = state.questions[index];
        state.currentIndex = index;
      }
    },
    selectPreviousQuestion: (state) => {
      const index = state.currentIndex - 1
      if(state.questions[index]) {
        state.current = state.questions[index];
        state.currentIndex = index;
      }
    },
    answerQuestion: (state, action: PayloadAction<AnswerType>) => {
      state.answers[action.payload.questionId] = action.payload;
    },
  
    resetTest: (state) => {
      state.current = undefined;
      state.currentIndex = 0;
      state.answers = {};
    },
  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      .addCase(fetchQuestionAsync.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchQuestionAsync.fulfilled, (state, action) => {
        state.status = 'idle';
        state.questions = action.payload;
        if(action.payload.length) {
          state.current = state.questions[0];
          state.currentIndex = 0;
        }
      })
      .addCase(fetchQuestionAsync.rejected, (state) => {
        state.status = 'failed';
      });
  },
});

export const { answerQuestion, resetTest, selectNextQuestion, selectPreviousQuestion } = personalityTestSlice.actions;

export default personalityTestSlice.reducer;
