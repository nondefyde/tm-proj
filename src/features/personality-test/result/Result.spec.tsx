import { fireEvent, render } from '@testing-library/react'
import { Result } from './index';
import * as react from 'react'
import * as hooks from '../../../app/hooks';
import { useAppSelector } from '../../../app/hooks';
import { resetTest } from '../PersonalityTestSlice';

const mockDispatch = jest.fn();
jest.mock('../../../app/hooks', () => ({
  useAppSelector: jest.fn(),
  useAppDispatch: () => mockDispatch
}));

describe('Result Component', () => {
  
  const useStateMock = jest.spyOn(react, 'useState')
  const useAppDispatchMock = jest.spyOn(hooks, 'useAppDispatch')
  const useAppSelectorMock = jest.spyOn(hooks, 'useAppSelector')
  
  const setShowAnswer = jest.fn();
  const showAnswer = false;
  
  afterEach(() => {
    useAppSelectorMock.mockClear();
    useAppDispatchMock.mockClear();
  });
  
  beforeEach(() => {
    useStateMock.mockImplementation(() => [showAnswer, setShowAnswer]);
    useAppDispatchMock.mockReturnValue(mockDispatch);
    useAppSelectorMock.mockImplementation((selectorFn: any) => selectorFn({
      personalityTest: {
        result: {},
        questions: [
          {
            id: '1',
            type: 'multipleSelection',
            title: 'Title',
            items: [
              {
                id: '1',
                identifier: 'A',
                description: 'Description'
              }
            ]
          },
        ],
        answers: {}
      },
    }));
  });
  
  it('should test snapshot', () => {
    const container = render(<Result/>)
    expect(container).toMatchSnapshot();
  });
  
  it('should take snapshot when showAnswer is true', () => {
    useStateMock.mockImplementation(() => [true, setShowAnswer]);
    const { getByText } = render(<Result/>)
    const showText = getByText(/Show my Answers/);
    expect(showText).toBeInTheDocument();
  });
  
  it('should call resetTest when retake button is click ', () => {
    const { getByTestId } = render(<Result/>)
    const retakeButton = getByTestId('retake');
    fireEvent.click(retakeButton);
    expect(mockDispatch).toHaveBeenCalledWith(resetTest());
  });

  it('should setShowAnswer when showAnswer button is clicked', () => {
    const { getByTestId } = render(<Result/>)
    const showAnswerButton = getByTestId('show-answer');
    fireEvent.click(showAnswerButton);
    expect(setShowAnswer).toHaveBeenCalledWith(!showAnswer);
  });
});
