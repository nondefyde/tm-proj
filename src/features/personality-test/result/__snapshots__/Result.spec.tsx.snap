// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result Component should test snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bcXHqe ciZuhd"
      >
        <div>
          <h4>
            Your Result
          </h4>
          <h2
            class="sc-gswNZR noKUc"
          />
          <div
            class="sc-dkrFOg gwmpyD"
          />
          <div
            class="sc-iBYQkv cbuvIf"
          >
            <button
              class="sc-gKPRtg kSQhHv"
              data-testid="retake"
            >
              Retake Question
            </button>
            <button
              class="sc-gKPRtg hWKZls"
              data-testid="show-answer"
            >
              Hide my Answers
            </button>
          </div>
        </div>
        <div
          class="sc-hLBbgP klOabs"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bcXHqe ciZuhd"
    >
      <div>
        <h4>
          Your Result
        </h4>
        <h2
          class="sc-gswNZR noKUc"
        />
        <div
          class="sc-dkrFOg gwmpyD"
        />
        <div
          class="sc-iBYQkv cbuvIf"
        >
          <button
            class="sc-gKPRtg kSQhHv"
            data-testid="retake"
          >
            Retake Question
          </button>
          <button
            class="sc-gKPRtg hWKZls"
            data-testid="show-answer"
          >
            Hide my Answers
          </button>
        </div>
      </div>
      <div
        class="sc-hLBbgP klOabs"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
